modules/order/src/MigrateExistingOrderProfiles.php

```diff
@@ -0,0 +1,123 @@
+   *
+   * Profiles that are associated with both shipping and billing information are
+   * handled differently. The original is migrated to be a billing profile. Then
+   * a copy is created that is migrated to be a shipping profile and the order
+   * shipments that use it are updated to use the new one.
+   *
+   * @param array $order_ids
+   *   An array of order IDs.
+   * @param \Drupal\commerce_order\Entity\OrderTypeInterface $order_type
+   *   The commerce_order_type entity.
+   * @param array $context
+   *   The batch context array.
+   */
+  public static function migrateProfiles(
+    array $order_ids,
+    OrderTypeInterface $order_type,
+    array &$context
+  ) {
+    $message = 'Migrating existing order profiles to use split profiles for
+      billing and shipping...';
+
+    $results = [];
+    foreach ($order_ids as $order_id) {
+      /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+      $order = \Drupal::entityTypeManager()
+        ->getStorage('commerce_order')
+        ->load($order_id);
+
+      // Grab the billing profile.
+      /** @var \Drupal\profile\Entity\ProfileInterface $billing_profile */
+      $billing_profile = $order->getBillingProfile();
+      if ($billing_profile) {
+        // Change the billing profile from the customer profile type to the
+        // customer_billing profile type.
+        $billing_profile->set('type', 'customer_billing');
+        $billing_profile->save();
+      }
+
+      // Grab the shipping profiles.
+      if ($order->shipments) {
+        foreach ($order->shipments->referencedEntities() as $shipment) {
+          /** @var \Drupal\profile\Entity\ProfileInterface $shipping_profile */
+          $shipping_profile = $shipment->getShippingProfile();
+
+          // We have distinct billing and shipping profile references.
+          if ($shipping_profile->id() != $billing_profile->id()) {
+            // Change the shipping profile from the customer profile type to the
+            // customer_shipping profile type.
+            $shipping_profile->set('type', 'customer_shipping');
+            $shipping_profile->save();
+          }
+          // Else, if the shipment references the same billing profile, create
+          // and save a new duplicate and reference this new shipping profile.
+          else {
+            /** @var \Drupal\profile\Entity\ProfileInterface $new_shipping_profile */
+            $new_shipping_profile = $shipping_profile->createDuplicate();
+
+            // Change the shipping profile from the customer profile type to the
+            // customer_shipping profile type.
+            $new_shipping_profile->set('type', 'customer_shipping');
+            $new_shipping_profile->save();
+
+            /** @var \Drupal\commerce_shipping\Entity\ShipmentInterface $shipment */
+            $shipment->setShippingProfile($new_shipping_profile);
+          }
+        }
+      }
```

Comment:
**OPTIONAL CHANGE REQUEST**

*Readability*

I'd refactor the billing/shipping processing parts into 2 separate protected functions that I would call from here. The length and complexity of the code is too much otherwise; it can simply be:

```
$billing_profile = $this->processBillingProfile($order)
$this->processShippingProfiles($order, $billing_profile)
````

Reply:
Yup, I agree.